### PR TITLE
add location service

### PR DESCRIPTION
**Pull request by StupidCodeFactory**

### CODE REVIEW

File: app/views/prison/visits/_prisoner_details.html.erb [hunk 1]
```diff
@@ -50,5 +50,11 @@
   <div class="column-one-quarter">
     <span class="text-secondary"><%= t('.prison_name') %></span>
     <div class="bold-small"><%= @visit.prison_name %></div>
+    <% case @visit.location_confirmed? %>
```
Comment (StupidCodeFactory) [inline app/views/prison/visits/_prisoner_details.html.erb]: @chrishanes that's where you should hook into

File: app/services/nomis/api.rb [hunk 2]
```diff
@@ -24,6 +24,19 @@ def initialize
       end
     end
 
+    def lookup_offender_location(offender_id:)
+      response = @pool.with do |client|
+        client.get("/offenders/#{offender_id}/location")
+      end
+    rescue Nomis::APIError => e
+      Raven.capture_exception(e, fingerprint: %w[nomis api_error])
+      PVB::Instrumentation.append_to_log(lookup_offender_location: false)
```
Comment (alan) [inline app/services/nomis/api.rb]: How about renaming the log key to `lookup_offender_location_error`?

File: app/services/nomis/api.rb [hunk 3]
```diff
@@ -24,6 +24,19 @@ def initialize
...
+      Nomis::Establishment.new
+    else
+      PVB::Instrumentation.append_to_log(lookup_offender_location: true)
+      Nomis::Establishment.new(response.fetch('establishment') { {} })
```
Comment (alan) [inline app/services/nomis/api.rb]: How come the fetch is needed? Does the api not return a response with an `establishment` key when is successful?

File: app/views/prison/visits/_prisoner_details.html.erb (hunk 1 above)
Comment (alan) [inline app/views/prison/visits/_prisoner_details.html.erb]: I think `location_confirmed` will need to be delegated in the visit decorator for this to work

File: app/services/nomis/api.rb [hunk 4]
```diff
@@ -24,6 +24,19 @@ def initialize
       end
     end
 
+    def lookup_offender_location(offender_id:)
+      response = @pool.with do |client|
+        client.get("/offenders/#{offender_id}/location")
+      end
+    rescue Nomis::APIError => e
+      Raven.capture_exception(e, fingerprint: %w[nomis api_error])
```
Comment (alan) [inline app/services/nomis/api.rb]: How about renaming this log key to `lookup_offender_location_error`?

File: app/services/nomis/api.rb (hunk 3 above)
Comment (alan) [inline app/services/nomis/api.rb]: Does the api not return a response with an `establishment` key when is successful? I think the fetch might be unnecessary.

File: app/views/prison/visits/_prisoner_details.html.erb (hunk 1 above)
Comment (alan) [inline app/views/prison/visits/_prisoner_details.html.erb]: I think this will need to be delegated in the visit decorator to work

File: app/models/prisoner_validation.rb [hunk 5]
```diff
@@ -1,6 +1,7 @@
 class PrisonerValidation
-  UNKNOWN            = 'unknown'.freeze
-  PRISONER_NOT_EXIST = 'prisoner_does_not_exist'.freeze
+  UNKNOWN                   = 'unknown'.freeze
+  PRISONER_NOT_EXIST        = 'prisoner_does_not_exist'.freeze
+  PRISONER_LOCATION_INVALID = 'prisoner_location_not_valid'.freeze
```
Comment (alan) [inline app/models/prisoner_validation.rb]: The validation for location needs the `unknown` state. For example the prisoner could be found and then the location api could timeout. All of the other checks follow the pattern of having 3 states: valid, invalid and unknown

File: app/models/prisoner_validation.rb [hunk 6]
```diff
@@ -9,18 +10,35 @@ def initialize(offender)
     self.offender = offender
   end
 
+  # rubocop:disable Metrics/MethodLength
   def active_offender_exists
-    unless Nomis::Api.enabled? && offender.api_call_successful?
```
Comment (alan) [inline app/models/prisoner_validation.rb]: Are we checking for the api being enabled somewhere else?

File: app/services/api_prisoner_checker.rb [hunk 7]
```diff
@@ -21,7 +14,28 @@ def error
...
+
+  def fetch_offender
+    Nomis::Api.instance.lookup_active_offender(
+      noms_id:       noms_id,
+      date_of_birth: date_of_birth
+    )
+  end
+
+  def null_offender
+    @null_offender ||= Nomis::NullOffender.new(api_call_successful: false)
```
Comment (alan) [inline app/services/api_prisoner_checker.rb]: No need to memoize since this will get called at most once.

Unrelated, but I wonder if the null offender could be a singleton object.

File: app/services/api_prisoner_checker.rb [hunk 8]
```diff
@@ -1,14 +1,7 @@
 class ApiPrisonerChecker
   def initialize(noms_id:, date_of_birth:)
-    @offender = if Nomis::Api.enabled? &&
```
Comment (alan) [inline app/services/api_prisoner_checker.rb]: Same here, is the api being disabled check gone?

File: app/services/api_prisoner_checker.rb [hunk 9]
```diff
@@ -1,14 +1,7 @@
 class ApiPrisonerChecker
```
Comment (alan) [inline app/services/api_prisoner_checker.rb]: Are the changes in this class mostly a refactor? It's confusing since this class is unrelated for the location service work.

File: app/services/slot_availability.rb [hunk 10]
```diff
@@ -29,7 +29,7 @@ def all_slots
 
 private
 
-  attr_reader :prison, :noms_id, :date_of_birth, :start_date, :end_date
```
Comment (alan) [inline app/services/slot_availability.rb]: How come these objects can't be value objects anymore? I don't see any need for them to be mutable.

File: app/services/slot_availability.rb [hunk 11]
```diff
@@ -76,8 +75,7 @@ def offender_availability_dates
   end
 
   def nomis_public_prisoner_availability_enabled?
-    Nomis::Api.enabled? &&
```
Comment (alan) [inline app/services/slot_availability.rb]: The api check is gone here also

File: app/services/staff_nomis_checker.rb [hunk 12]
```diff
@@ -6,8 +6,7 @@ class StaffNomisChecker
   NOT_LIVE = 'not_live'.freeze
 
   def initialize(visit)
-    @visit = visit
-    @nomis_api_enabled = Nomis::Api.enabled?
```
Comment (alan) [inline app/services/staff_nomis_checker.rb]: This class also removes the check for the nomis api

File: app/views/prison/visits/_prisoner_details.html.erb [hunk 13]
```diff
@@ -50,5 +50,10 @@
   <div class="column-one-quarter">
     <span class="text-secondary"><%= t('.prison_name') %></span>
     <div class="bold-small"><%= @visit.prison_name %></div>
-  </div>
+    <% if @visit.location_verified? %>
```
Comment (alan) [inline app/views/prison/visits/_prisoner_details.html.erb]: It's missing the third state `unknown`

File: spec/controllers/api/validations_controller_spec.rb [hunk 14]
```diff
@@ -12,16 +12,16 @@
   end
 
   describe 'visitors' do
-    let!(:prison) { FactoryGirl.create(:prison) }
+    let!(:prison)   { create(:prison) }
```
Comment (alan) [inline spec/controllers/api/validations_controller_spec.rb]: Why are these changes in this PR? How is it related to the location service?

File: app/services/slot_availability.rb (hunk 10 above)
Comment (StupidCodeFactory) [inline app/services/slot_availability.rb]: It's not changed, this change just prevent(ish) accessing the instance variable directly and encourages using accessor methods (initialize method).
I don't mind reverting it, in my opinion it's safer to mutate instance variables to through accessor methods.

File: app/services/api_prisoner_checker.rb (hunk 7 above)
Comment (StupidCodeFactory) [inline app/services/api_prisoner_checker.rb]: if need to be able to have a different id

File: app/services/slot_availability.rb (hunk 10 above)
Comment (alan) [inline app/services/slot_availability.rb]: Ah, yes, I missed it was private. Ignore my comment.

File: app/services/api_prisoner_checker.rb (hunk 7 above)
Comment (StupidCodeFactory) [inline app/services/api_prisoner_checker.rb]: this does not really hurt and would be the same as accessing the instance variable directly which we do all over the place

File: app/models/prisoner_validation.rb [hunk 15]
```diff
@@ -1,14 +1,23 @@
 class PrisonerValidation
-  UNKNOWN            = 'unknown'.freeze
-  PRISONER_NOT_EXIST = 'prisoner_does_not_exist'.freeze
+  VALID                     = 'ok'.freeze
```
Comment (alan) [inline app/models/prisoner_validation.rb]: This is not used I think

File: app/models/prisoner_validation.rb [hunk 16]
```diff
@@ -17,10 +26,30 @@ def active_offender_exists
 
     unless offender.valid?
       errors.add :base, PRISONER_NOT_EXIST
+      return
+    end
+
+    unless prisoner_location.api_call_successful?
```
Comment (alan) [inline app/models/prisoner_validation.rb]: I think here we need to check for the offender being valid before calling the prisoner_location

File: app/services/staff_nomis_checker.rb [hunk 17]
```diff
@@ -14,10 +18,10 @@ def prisoner_existance_status
     return NOT_LIVE unless prisoner_check_enabled?
 
     case prisoner_existance_error
-    when nil
-      VALID
-    when UNKNOWN
-      UNKNOWN
+    when nil              then VALID
```
Comment (alan) [inline app/services/staff_nomis_checker.rb]: I find this confusing; it mixes the prisoner existence with location and depends on the first sorted error message.

The location is almost the same as the prisoner availability in terms of how it interacts with prisoner existence:

- If prisoner existence is invalid then location is unknown: The user could have typoed the noms_id but still be valid (and existence be invalid because the date of birth does not match). If we lookup the location then it will be the location of a completely different prisoner.

- If the prisoner existence is unknown then the location is unknown. Similar to above but the existence api times out, responds with a non 200 code.

- Prisoner existence is valid, then location could be valid, unknown or invalid

File: app/views/prison/visits/_prisoner_details.html.erb [hunk 18]
```diff
@@ -7,21 +7,23 @@
...
+<% when StaffNomisChecker::UNKNOWN, StaffNomisChecker::LOCATION_UNKNOWN %>
   <div class="notice push-top">
     <i class="icon icon-important">
       <span class="visuallyhidden"><%= t('warning', scope: :shared) %></span>
     </i>
     <strong class="bold-small"><%= t(".#{@visit.prisoner_existance_error}_text") %></strong>
   </div>
 <% end %>
+
+
```
Comment (alan) [inline app/views/prison/visits/_prisoner_details.html.erb]: ✂️ ?
